compiler-rt/lib/fuzzer/FuzzerMutate.cpp

```diff
@@ -171,21 +171,27 @@ size_t MutationDispatcher::ApplyDictionaryEntry(uint8_t *Data, size_t Size,
                                                 size_t MaxSize,
                                                 DictionaryEntry &DE) {
   const Word &W = DE.GetW();
-  bool UsePositionHint = DE.HasPositionHint() &&
-                         DE.GetPositionHint() + W.size() < Size &&
-                         Rand.RandBool();
-  if (Rand.RandBool()) {  // Insert W.
-    if (Size + W.size() > MaxSize) return 0;
-    size_t Idx = UsePositionHint ? DE.GetPositionHint() : Rand(Size + 1);
-    memmove(Data + Idx + W.size(), Data + Idx, Size - Idx);
-    memcpy(Data + Idx, W.data(), W.size());
-    Size += W.size();
-  } else {  // Overwrite some bytes with W.
-    if (W.size() > Size) return 0;
-    size_t Idx = UsePositionHint ? DE.GetPositionHint() : Rand(Size - W.size());
-    memcpy(Data + Idx, W.data(), W.size());
-  }
-  return Size;
+  bool UsePositionHint =
+      DE.HasPositionHint() &&
+      DE.GetPositionHint().Idx + DE.GetPositionHint().Size < Size &&
+      Rand.RandBool();
+  // Either insert W or overwrite some bytes with it using a size hint if
+  // possible.
+  size_t HintSize =
+      Rand.RandBool()
+          ? 0
+          : (UsePositionHint ? DE.GetPositionHint().Size : W.size());
+  if (HintSize > Size)
+    return 0;
+  size_t Idx =
+      UsePositionHint ? DE.GetPositionHint().Idx : Rand(Size + 1 - HintSize);
+  if (Size + W.size() - HintSize > MaxSize)
+    return 0;
+  // Overwrite HintSize bytes at Idx with W (HintSize == 0 corresponds to
+  // insertion).
+  memmove(Data + Idx + W.size(), Data + Idx + HintSize, Size - Idx - HintSize);
+  memcpy(Data + Idx, W.data(), W.size());
+  return Size + W.size() - HintSize;
```

Comment:
I find it a bit tricky to verify that the new code is more or less equivalent to the old one (modulo the size hint). If I understand correctly, the old code flips a coin to either insert the word or overwrite some part of the data with it, with the index being either an arbitrary position or the one specified by the hint, depending again on randomness.

The new code somehow adds the hinted size to the logic, but I can't fully figure out how, especially because it always writes `W.size()` bytes. Is the hinted size greater or smaller than the word size, and how does it affect the `memmove`?

Reply:
The idea is that insertion is equivalent to overwriting 0 bytes. `HintSize` has a 50% chance of being 0, meaning that `W` will be inserted, and a 50% chance of being the hint size (which is taken to be `W.size()` if there is no hint, which recovers the overwrite path of the old code).

The hinted size can be greater or smaller than the word size and the memmove should ensure that the `HintSize` bytes at position `Idx` are replaced with the `W.size()` byte of `W`.

Reply:
I see, thanks!